### PR TITLE
feat: send all SmartContractEvents to DDC

### PR DESCRIPTION
**Pull request by vadim-hleif**
@spoluyan I have doubts about pub key for batch transfer. As I remember we want to get all NFT related actions, so, maybe it's better to split batch event into multiple single events? 
Is the fact that some transfer happened in batch important?

### CODE REVIEW

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt [hunk 1]
```diff
@@ -5,7 +5,7 @@ import java.math.BigInteger
 sealed class SmartContractEvent(
     open val blockSignedAt: String,
     open val txHash: String
-)
+) : DdcStorableData
```
Comment (spoluyan) [inline src/main/kotlin/nft/freeport/event/SmartContractEvent.kt:8]: Not all SC events are DdcStorable, only NFT-related.

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt [hunk 2]
```diff
@@ -25,15 +27,19 @@ data class TransferBatch(
     val to: String,
     val ids: List<String>,
     val amounts: List<BigInteger>
-) : SmartContractEvent(blockSignedAt, txHash)
+) : SmartContractEvent(blockSignedAt, txHash) {
+    override fun userPubKey(): String = operator
```
Comment (spoluyan) [inline src/main/kotlin/nft/freeport/event/SmartContractEvent.kt:31]: Operator makes no sense in DDC - it's SC address. We should split batch to single events and store them individually.

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt [hunk 3]
```diff
@@ -25,15 +27,19 @@ data class TransferBatch(
...
 
 data class JointAccountShareCreated(
     override val blockSignedAt: String,
     override val txHash: String,
     val account: String,
     val owner: String,
     val fraction: Int
-) : SmartContractEvent(blockSignedAt, txHash)
+) : SmartContractEvent(blockSignedAt, txHash) {
+    override fun userPubKey(): String = account
```
Comment (spoluyan) [inline src/main/kotlin/nft/freeport/event/SmartContractEvent.kt:41]: No need to store in DDC.

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt [hunk 4]
```diff
@@ -63,13 +73,17 @@ data class TakeOffer(
...
+    override fun userPubKey(): String = nftId
+}
 
 data class SetExchangeRate(
     override val blockSignedAt: String,
     override val txHash: String,
     val cereUnitsPerPenny: BigInteger
-) : SmartContractEvent(blockSignedAt, txHash)
+) : SmartContractEvent(blockSignedAt, txHash) {
+    override fun userPubKey(): String = txHash
```
Comment (spoluyan) [inline src/main/kotlin/nft/freeport/event/SmartContractEvent.kt:85]: No need to store in DDC.

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt (hunk 1 above)
Comment (vadim-hleif) [reply src/main/kotlin/nft/freeport/event/SmartContractEvent.kt]: done

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt (hunk 3 above)
Comment (vadim-hleif) [reply src/main/kotlin/nft/freeport/event/SmartContractEvent.kt]: done

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt (hunk 4 above)
Comment (vadim-hleif) [reply src/main/kotlin/nft/freeport/event/SmartContractEvent.kt]: done

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt (hunk 3 above)
Comment (vadim-hleif) [reply src/main/kotlin/nft/freeport/event/SmartContractEvent.kt]: done

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt (hunk 2 above)
Comment (vadim-hleif) [reply src/main/kotlin/nft/freeport/event/SmartContractEvent.kt]: done

**Comment by vadim-hleif** [comment]
@spoluyan please also take a look on changing object props description to with(){} construction 

I haven't faced this problem before, but now I've realised that it can cause bugs in case of reordering
tests event didn't fail, there were just idea warning which can be easily missed

File: src/main/kotlin/nft/freeport/event/SmartContractEvent.kt [hunk 5]
```diff
@@ -25,7 +33,7 @@ data class TransferBatch(
     val to: String,
     val ids: List<String>,
     val amounts: List<BigInteger>
-) : SmartContractEvent(blockSignedAt, txHash)
+) : NftRelatedEvent(blockSignedAt, txHash, from)
```
Comment (vadim-hleif) [inline src/main/kotlin/nft/freeport/event/SmartContractEvent.kt:36]: I am not sure which field I should pass here @spoluyan 
decided that from is the best option, please assist

Comment (vadim-hleif) [reply src/main/kotlin/nft/freeport/event/SmartContractEvent.kt]: discussed it in slack

batch should not be sent to DDC, mark it just as SmartContractEvent

**Review by upalinski — APPROVED** [review]

**Comment by spoluyan** [comment]
Merged to refactoring branch

**Pull request closed without merging**